sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -675,7 +675,7 @@ case class AlterTableRecoverPartitionsCommand(
     // This is always the case for Hive format tables, but is not true for Datasource tables created
     // before Spark 2.1 unless they are converted via `msck repair table`.
     spark.sessionState.catalog.alterTable(table.copy(tracksPartitionsInCatalog = true))
-    catalog.refreshTable(tableName)
+    spark.catalog.refreshTable(tableIdentWithDB)
```

Comment:
Here, we invoke https://github.com/apache/spark/blob/3fdbc48373cdf12b8ba05632bc65ad49b7af1afb/sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala#L514 instead of https://github.com/apache/spark/blob/cc1d9d25fb4c2e4af912d6f9802de8f351c32deb/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala#L990 . The last one doesn't refresh the cache.

Reply:
@cloud-fan I don't know why we have 2 implementations of `refreshTable()`. I would merge them and call one from another.

Reply:
yea we should unify these two refreshTable

Reply:
+1 for the unification.

Reply:
I think the former calls the latter which only refreshes metadata cache and is used in multiple other places.

Reply:
I wonder what are the use cases when we need to update only meta-data but not cached table data. Looking at the places where `SessionCatalog.refreshTable` is used:
1. https://github.com/apache/spark/blob/c62b84a0432e51fd10e628088ee311dc3be73d2f/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala#L113
2. https://github.com/apache/spark/blob/271c4f6e00b7bc7c47d84a8e59018e84a19c9822/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L729
3. https://github.com/apache/spark/blob/2ab77d634f2e87b080786f4f39cb17e0994bc550/sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala#L245
4. https://github.com/apache/spark/blob/ddc0d5148ac6decde160cca847b5db5d6de1be58/sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala#L233
5. https://github.com/apache/spark/blob/ddc0d5148ac6decde160cca847b5db5d6de1be58/sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala#L394

In all those ^^ places, updating of cached table data makes sense, IMHO.

Reply:
I think there are cases like `alterTableStats` (may be the only one?) which only trigger metadata change. In the above, some already uncache the table data, although via different paths such as `CommandUtils.uncacheTableOrView` or `cacheManager.uncacheQuery`. Also, the other `refreshTable` recaches the target table but sometimes seems we just want to remove the cache but still refreshes the metadata.

Of course, it will be very helpful if we can simplify the code a bit. It also seems there are still quite a few cases where cache is not properly handled, both in v1 and v2.